### PR TITLE
sink (ticdc): add glue schema registry

### PR DESCRIPTION
**Pull request by asddongmen**
### What problem does this PR solve?

Issue Number: ref #9467 

### What is changed and how it works?
**Why is this needed?**
Some users want to use AWS Glue Schema Registry as their schema registry when using Avro protocol.

**What is changed?**
This pull request implements an AWS Glue Schema Registry manager, which allows TiCDC to use the Glue Schema Registry when replicating messages using the Avro protocol.

**How does it works?**
A `glue-schema-registry-config` into the `sink.kafka-config`, you need to set it when you want to use glue schema registry.
To configure `glue-schema-registry-config` into `sink.kafka-config`, use the following example:
```toml
[sink]
[sink.kafka-config.glue-schema-registry-config]
region="us-west-1"  
registry-name="ticdc-test"
access-key="xxxx"
secret-access-key="xxxx"
tokne="xxxx"
```
`region` and `registry-name` are required, while `access-key`, `secret-access-key`, and `token` are optional. The best practice is to set your AWS credentials in environment variables or store them in the `~/.aws/credentials` file other than set them in the changefeed's config file. For more information, refer to the [AWS documentation](https://aws.github.io/aws-sdk-go-v2/docs/configuring-sdk/#specifying-credentials).

#### Tests 

 - Unit test
 - Manual test (add detailed scripts or steps below)

1. prepare changefeed config file
```toml
[sink]
[sink.kafka-config.glue-schema-registry-config]
region="us-west-1"
registry-name="ticdc-test"
access-key="xxxx"
secret-access-key="xxxx"
```
3. create a changefeed.
```
./cdc cli changefeed create --server=127.0.0.1:8300 --changefeed-id="kafka-avro" --sink-uri="kafka://xxxx" --config changefeed_glue.toml
```
4. Insert some data into upstream TiDB, do some alter table DDLs.
5. Run a kafka consumer and  use AWS official python deserializer to decode the message.
```
{'a': 7} {"type": "record", "name": "t3", "namespace": "default.a", "fields": [{"name": "a", "type": {"type": "int", "connect.parameters": {"tidb_type": "INT"}}}]}
{'a': 8, 'name': 'dongmen'} {"type": "record", "name": "t3", "namespace": "default.a", "fields": [{"name": "a", "type": {"type": "int", "connect.parameters": {"tidb_type": "INT"}}}, {"default": null, "name": "name", "type": ["null", {"type": "string", "connect.parameters": {"tidb_type": "TEXT"}}]}]}
{'a': 9, 'name': 'dongmen'} {"type": "record", "name": "t3", "namespace": "default.a", "fields": [{"name": "a", "type": {"type": "int", "connect.parameters": {"tidb_type": "INT"}}}, {"default": null, "name": "name", "type": ["null", {"type": "string", "connect.parameters": {"tidb_type": "TEXT"}}]}]}
{'a': 10, 'name': 'guagua', 'email': 'xxx'} {"type": "record", "name": "t3", "namespace": "default.a", "fields": [{"name": "a", "type": {"type": "int", "connect.parameters": {"tidb_type": "INT"}}}, {"default": null, "name": "name", "type": ["null", {"type": "string", "connect.parameters": {"tidb_type": "TEXT"}}]}, {"default": null, "name": "email", "type": ["null", {"type": "string", "connect.parameters": {"tidb_type": "TEXT"}}]}]}
{'a': 12, 'name': 'drop email'} {"type": "record", "name": "t3", "namespace": "default.a", "fields": [{"name": "a", "type": {"type": "int", "connect.parameters": {"tidb_type": "INT"}}}, {"default": null, "name": "name", "type": ["null", {"type": "string", "connect.parameters": {"tidb_type": "TEXT"}}]}]}
```
6.  All data received and decoded are as expected.

### Release note 

```release-note
support glue schema registry.
```

### CODE REVIEW

File: pkg/sink/codec/avro/avro.go [hunk 1]
```diff
@@ -1015,9 +1008,18 @@ func NewBatchEncoderBuilder(ctx context.Context,
 	changefeedID model.ChangeFeedID,
 	config *common.Config,
 ) (codec.RowEventEncoderBuilder, error) {
-	schemaM, err := NewAvroSchemaManager(ctx, config.AvroSchemaRegistry, nil)
-	if err != nil {
-		return nil, errors.Trace(err)
+	var schemaM SchemaManager
+	var err error
+	if config.AvroConfluentSchemaRegistry != "" {
```
Comment (nongfushanquan) [inline pkg/sink/codec/avro/avro.go:1013]: is  there still other schema registry?

Comment (asddongmen) [reply pkg/sink/codec/avro/avro.go:1013]: There are three main Schema Registry options :Confluent Schema Registry, Aws Glue, Apicurio.

Now TiCDC only support Confluent Schema Registry.

Comment (asddongmen) [reply pkg/sink/codec/avro/avro.go:1013]: This check here is to decide whether the decoder should use Confluent Schema Registry or Aws Glue.

File: cdc/api/v2/model.go [hunk 2]
```diff
@@ -770,24 +792,25 @@ type Table struct {
...
+	EnablePartitionSeparator         *bool                     `json:"enable_partition_separator,omitempty"`
+	FileIndexWidth                   *int                      `json:"file_index_width,omitempty"`
+	EnableKafkaSinkV2                *bool                     `json:"enable_kafka_sink_v2,omitempty"`
+	OnlyOutputUpdatedColumns         *bool                     `json:"only_output_updated_columns,omitempty"`
+	DeleteOnlyOutputHandleKeyColumns *bool                     `json:"delete_only_output_handle_key_columns"`
+	SafeMode                         *bool                     `json:"safe_mode,omitempty"`
+	KafkaConfig                      *KafkaConfig              `json:"kafka_config,omitempty"`
+	MySQLConfig                      *MySQLConfig              `json:"mysql_config,omitempty"`
+	CloudStorageConfig               *CloudStorageConfig       `json:"cloud_storage_config,omitempty"`
+	GlueSchemaRegistryConfig         *GlueSchemaRegistryConfig `toml:"glue-schema-registry-config" json:"glue-schema-registry-config,omitempty"`
```
Comment (sdojjy) [inline cdc/api/v2/model.go:813]: api don't need `toml`

File: cdc/api/v2/model.go [hunk 3]
```diff
@@ -1055,3 +1078,15 @@ type ChangefeedStatus struct {
...
+
+type GlueSchemaRegistryConfig struct {
+	// Name of the schema registry
+	RegistryName string `toml:"registry-name" json:"registry-name"`
+	// Region of the schema registry
+	Region string `toml:"region" json:"region"`
+	// AccessKeyID of the schema registry
+	AccessKeyID string `toml: "access-key-id" json:"access-key-id,omitempty"`
+	// AccessKeySecret of the schema registry
+	AccessKeySecret string `toml: "access-key-secret" json:"access-key-secret,omitempty"`
```
Comment (sdojjy) [inline cdc/api/v2/model.go:1090]: use underscore to combine words

File: pkg/sink/codec/avro/avro.go [hunk 4]
```diff
@@ -70,7 +70,7 @@ type avroEncodeResult struct {
 	data []byte
 	// schemaID is encoded into the avro message, consumer should use this to fetch the schema.
```
Comment (sdojjy) [inline pkg/sink/codec/avro/avro.go:71]: update the comment

File: pkg/sink/codec/avro/schema_manager.go [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+}
+
+// SchemaGenerator represents a function that returns an Avro schema in JSON.
+// Used for lazy evaluation
+type SchemaGenerator func() (string, error)
+
+type schemaID struct {
+	// cID is the Confluent Schema ID, it represents
+	// a unique schema in Confluent Schema Registry
+	cID int
```
Comment (sdojjy) [inline pkg/sink/codec/avro/schema_manager.go:25]: it's better to use the full name?

File: cdc/api/v2/model.go [hunk 6]
```diff
@@ -413,6 +413,16 @@ func (c *ReplicaConfig) toInternalReplicaConfigWithOriginConfig(
 			res.Sink.AdvanceTimeoutInSec = util.AddressOf(*c.Sink.AdvanceTimeoutInSec)
 		}
 
+		if c.Sink.GlueSchemaRegistryConfig != nil {
```
Comment (3AceShowHand) [inline cdc/api/v2/model.go:416]: It loos this config only belongs to the kafka sink ? so put it inside the `kafkaConfig` ?

File: cdc/api/v2/model.go [hunk 7]
```diff
@@ -413,6 +413,16 @@ func (c *ReplicaConfig) toInternalReplicaConfigWithOriginConfig(
...
+			res.Sink.GlueSchemaRegistryConfig = &config.GlueSchemaRegistryConfig{
+				RegistryName:    c.Sink.GlueSchemaRegistryConfig.RegistryName,
+				Region:          c.Sink.GlueSchemaRegistryConfig.Region,
+				AccessKey:       c.Sink.GlueSchemaRegistryConfig.AccessKey,
```
Comment (3AceShowHand) [inline cdc/api/v2/model.go:420]: `AccessKey`, `SecretAccessKey`, `Token` should be masked to prevent leak ?

File: go.mod [hunk 8]
```diff
@@ -12,6 +12,9 @@ require (
 	github.com/VividCortex/mysqlerr v1.0.0
 	github.com/apache/pulsar-client-go v0.11.0
 	github.com/aws/aws-sdk-go v1.44.259
+	github.com/aws/aws-sdk-go-v2 v1.19.1
```
Comment (3AceShowHand) [inline go.mod:15]: Why do we use `aws-sdk-go` v1 and v2 at the same time ? shall we only keep v2 ?

File: pkg/sink/codec/avro/avro.go [hunk 9]
```diff
@@ -68,9 +68,8 @@ func (r *avroEncodeInput) Swap(i, j int) {
 
 type avroEncodeResult struct {
 	data []byte
-	// schemaID is encoded into the avro message, consumer should use this to fetch the schema.
-	// it's the global schema id for all schema in the registry.
-	schemaID int
+	// header is the message header
+	header []byte
```
Comment (3AceShowHand) [inline pkg/sink/codec/avro/avro.go:72]: please add more comment about the header to explain it's usage

File: pkg/sink/codec/avro/avro.go [hunk 10]
```diff
@@ -1015,9 +1007,22 @@ func NewBatchEncoderBuilder(ctx context.Context,
 	changefeedID model.ChangeFeedID,
 	config *common.Config,
 ) (codec.RowEventEncoderBuilder, error) {
-	schemaM, err := NewAvroSchemaManager(ctx, config.AvroSchemaRegistry, nil)
-	if err != nil {
-		return nil, errors.Trace(err)
+	var schemaM SchemaManager
+	var err error
+
+	switch config.SchemaRegistryType() {
```
Comment (3AceShowHand) [inline pkg/sink/codec/avro/avro.go:1013]: `config.SchemaRegistryType()` can be called only once here

**Review by 3AceShowHand — APPROVED** [review]

File: go.mod (hunk 8 above)
Comment (asddongmen) [reply go.mod:15]: BR package is using v1.

File: cdc/api/v2/model.go (hunk 6 above)
Comment (asddongmen) [reply cdc/api/v2/model.go]: done

File: pkg/config/replica_config.go [hunk 11]
```diff
@@ -206,6 +206,19 @@ func (c *ReplicaConfig) ValidateAndAdjust(sinkURI *url.URL) error { // check sin
 		if err != nil {
 			return err
 		}
+
+		if c.Sink.SchemaRegistry != nil &&
```
Comment (CharlesCheung96) [inline pkg/config/replica_config.go:210]: Better to move such validation to sink.ValidateAndAjust

**Review by CharlesCheung96 — APPROVED** [review]

**Review by nongfushanquan — APPROVED** [review]

**Pull request merged**